Fix arrayopt checks

fwc opened the pull request:
<!--
Thank you for contributing to KLEE. We are looking forward to reviewing your PR. However, given the small number of active reviewers and our limited time, it might take a while to do so. We aim to get back to each PR within one month, and often do so within one week.

To review your PR, please add a summary of the proposed changes and ensure all items are fulfilled in the checklist above, by placing an "x" inside each applicable pair of brackets. More details about each item can be found in the [Developer's Guide](https://klee.github.io/docs/developers-guide/#pull-requests).
-->

## Summary: 

Some tests in the `ArrayOpt` folder were missing `FileCheck` invocations to check for the default `CHECK` directive.
This PR adds those invocations and ~~suggests~~ fixes ~~for most of~~ the ensuing failures.

Two tests (`test-mix.c` and `test_cache.c`) are not completely fixed since the `"Good"` branch is unreachable. If the condition is changed to make that branch reachable (by replacing `6` with `5`), the optimizer is unable to optimize the expression, which results in the test failing.

```c
  klee_make_symbolic(&k, sizeof(k), "k");
  klee_assume(k < 5);
  klee_assume(k >= 0);
  // ...
  if (array[k] + k - 2 == 6)
    printf("Good\n"); // unreachable
```

## Checklist:
- [x] The PR addresses a single issue.  If it can be divided into multiple independent PRs, please do so.
- [x] The PR is divided into a logical sequence of commits OR a single commit is sufficient.
- [x] There are no unnecessary commits (e.g. commits fixing issues in a previous commit in the same PR).
- [x] Each commit has a meaningful message documenting what it does.
- [x] All messages added to the codebase, all comments, as well as commit messages are spellchecked.
- [x] The code is commented OR not applicable/necessary.
- [x] The patch is formatted via clang-format OR not applicable (if explicitly overridden leave unchecked and explain).
- [x] There are test cases for the code you added or modified OR no such test cases are required.

251 commented:
Hi,

you're replacing the pattern:
```
// RUN: %klee --write-kqueries --output-dir=%t.klee-out --optimize-array=value %t.bc > %t.log 2>&1
// RUN: FileCheck %s -input-file=%t.log -check-prefix=CHECK-OPT_V
```
with:
```
// RUN: %klee --write-kqueries --output-dir=%t.klee-out --optimize-array=value %t.bc > %t.out 2> %t.log
// RUN: cat %t.out %t.log | FileCheck %s
// RUN: FileCheck %s -input-file=%t.log -check-prefix=CHECK-OPT_V
```
what seems redundant to me. In the original test the outputs of stdout/stderr are redirected into `%t.log` and than handed over to `FileCheck`. In your version you redirect the output into two files, cat them together and pipe them into FileCheck (+ the original call is still in there with a log file that only has half the content now).

fwc commented:
I added the _'concatenating two files'_-indirection to ensure that `stdout` is read before `stderr` by `FileCheck`. If this would not be the case, tests would fail since `YES` and `KLEE done: ...` would be switched up. (c.f. [commit msg](https://github.com/klee/klee/pull/1563/commits/4a885a659561465d27e6289a8e3ed0b91f9dc3e5)).

Merging both `FileCheck` invocations should look like this:

```
// RUN: %klee --write-kqueries --output-dir=%t.klee-out --optimize-array=value %t.bc > %t.out 2> %t.err
// RUN: cat %t.out %t.err | FileCheck %s -check-prefix=CHECK-OPT_V -check-prefix=CHECK
```

I think I did decide against this approach to keep the diff as minimal as possible.

If you wish I can update the PR to this (or another) approach.

251 commented:
If the order really is an issue wouldn't [CHECK-DAG](https://llvm.org/docs/CommandGuide/FileCheck.html#the-check-dag-directive) be the easier fix? (I think it's `CHECK-LABEL-DAG` instead of `CHECK-DAG-LABEL`.)

fwc commented:
Yes, the unlabelled CHECKs where not checked ("accidentally disabled"), which allowed the breakages fixed in https://github.com/klee/klee/pull/1563/commits/fd5691bb38431199525de2b58cec2d8e12db4606.

To be honest, I don't remember if I tried `CHECK-DAG` and was either unable to fix it that way or if I was unsatisfied with the result (I had to change the `FileCheck` invocations anyway...) or if I didn't even thought about it.

ccadar commented:
@fwc are you willing to try `CHECK-DAG` again?  I agree that this would be the better way to do this.

Also commits 2 & 3 need to be squashed in the previous ones. 

Commit 4 (test/ArrayOpt mix, cache: make test path reachable) could be submitted as a different PR, as it's conceptually different.

fwc commented:
@251 @ccadar I reworked the PR to use `CHECK-DAG`, squashed the first three commits and removed the latter two involving the hidden optimizer failure.

ccadar commented:
Thanks @fwc this one is ready to be merged. 
@251 any other comments?
